code/datums/components/personalshield.dm

```diff
@@ -0,0 +1,148 @@
+/datum/component/personalshield
+	var/current_charges = 3
+	/// How many charges total the shielding has
+	var/max_charges = 3
+	/// How long after we've been shot before we can start recharging.
+	var/recharge_delay = 20 SECONDS
+	/// How often recharge_rate charges are added when recharging
+	var/recharge_speed = 2 SECONDS
+	/// How many charges it gains per recharge cycle
+	var/recharge_rate = 1
+	/// for clothing this requires it to be equiped, for non-clothing it must be in-hand
+	var/must_be_worn_or_held = FALSE
+	/// timerid of recharging
+	var/recharge_timer
+	/// the current wearer
+	var/mob/living/carbon/wearer
+	/// icon state of activated shield overlay
+	var/shield_on = "shield-old"
+	/// items that can recharge this
+	var/list/item_rechargable
+	/// icon_states that can be toggled between
+	var/list/togglable_states
+	var/toggle_state = 1
+
+/datum/component/personalshield/Initialize(max_charges = 3, current_charges, recharge_delay = 20 SECONDS, recharge_speed = 2 SECONDS, recharge_rate = 1, must_be_worn_or_held = FALSE, shield_on = "shield-old", list/item_rechargable, list/togglable_states)
+	if(!istype(parent, /obj/item))
+		return COMPONENT_INCOMPATIBLE
+	src.max_charges = max_charges
+	if(!current_charges)
+		src.current_charges = max_charges
+	else
+		src.current_charges = current_charges
+	src.recharge_delay = recharge_delay
+	src.recharge_speed = recharge_speed
+	src.recharge_rate = recharge_rate
+	src.must_be_worn_or_held = must_be_worn_or_held
+	src.shield_on = shield_on
+	if(item_rechargable)
+		src.item_rechargable = typecacheof(item_rechargable)
+	src.togglable_states = togglable_states
+
+/datum/component/personalshield/RegisterWithParent()
+	. = ..()
+	RegisterSignal(parent, COMSIG_ITEM_EQUIPPED, .proc/equipped)
+	RegisterSignal(parent, COMSIG_ITEM_AFTER_PICKUP, .proc/pickup)
+	RegisterSignal(parent, COMSIG_ITEM_DROPPED, .proc/dropped)
+	RegisterSignal(parent, COMPONENT_SHIELD_TOGGLE_COLOR, .proc/toggle_shield_color)
+
+/datum/component/personalshield/UnregisterFromParent()
+	. = ..()
+	UnregisterSignal(parent, list(
+		COMSIG_ITEM_EQUIPPED, 
+		COMSIG_ITEM_AFTER_PICKUP, 
+		COMSIG_ITEM_DROPPED,
+		COMPONENT_SHIELD_TOGGLE_COLOR))

```

Comment:
I recommend calling this here. Otherwise component deletion isn't properly deleting its references and also not removing the overlay.
```
if(wearer)
	dropped(null, wearer)
```